### PR TITLE
Searching (WIP)

### PR DESCRIPTION
**Pull request by Mrmaxmeier**
:construction: WIP

To-dos
- [ ] ui and ux that doesn't suck
  - [x] activity indicator
  - [x] empty state
- [x] gc database data
  - [x] after switching tabs
- [ ] implement pagination
- [x] hide search behind env var `ENABLE_SEARCH=1`
- [x] battle the bitrot

### CODE REVIEW

**Comment by fasterthanlime** [comment]
@Mrmaxmeier I recently changed how the game-store passes data down to the app-store (since 0.11.0), so you might want to take a look at that first

**Comment by fasterthanlime** [comment]
lemme get the icomoon one for you

**Comment by fasterthanlime** [comment]
@Mrmaxmeier re throttling API calls, it's already handled in `util/api.js`, take a look near the end of the file, ctrl-f `cooldown`

**Comment by Mrmaxmeier** [comment]
@fasterthanlime Can you take a look at the current state of this PR?

Pagination isn't ready yet, because the API doesn't return the [total amount](https://github.com/itchio/itch/pull/231/files#diff-049047752b1fcc4ac913a85d04af3de9R170) of search results.
I personally think the search bar doesn't quite fit in. The loading dimmer isn't really elegant either.
Let me know if there's something that I should change.

anim

**Comment by fasterthanlime** [comment]
Ah Max, the interaction looks neat! I agree that the UI needs some work
(and I wonder if the search bar shouldn't always be there?)

Anyway I'm currently focused on i18n+preferences with matias, but will take
a look at this soon

**Comment by fasterthanlime** [comment]
I've been thinking about this a lot lately and the main issue for me is that we can't cache all search results in the db - it'll get too big. We need a way to mark these game records as "transient" and be able to garbage collect them when we don't need them anymore.

I haven't yet found a reliable way to do that, there's a lot that can happen. For example let's say we had a `_fetched_for` field in our game db records, that contained a tab name such as "search/goblins".

Whenever we garbage collect the DB, we could see if the tab is still open and remove records whose tabs have been closed - but what if the game has been installed since? Or added to a collection? We need a way to check that too.

Maybe the DB garbage collection routine should first collect all game IDs from collections, owned, etc., so that it knows what not to delete - and then we don't even need `_fetched_for`

File: app/stores/game-store.js [hunk 1]
```diff
@@ -148,6 +149,55 @@ async function fetch_keys (type, page) {
...
+  game_ids = game_ids.concat(_.pluck(res.games, 'id'))
+  for (let game of res.games) {
+    games[game.id] = game
+  }
+
+  await db.save_games(res.games)
+  AppActions.search_fetched(query, game_ids, games)
+
+  if (fetched < total_items) {
+    await fetch_search_page(query, page + 1, games)
```
Comment (fasterthanlime) [inline app/stores/game-store.js]: Hmm, we probably shouldn't fetch all search pages in one go - as opposed to collections which we want to have available offline

**Comment by fasterthanlime** [comment]
I want to bring this PR and #328 into master so it doesn't bitrot again, but hide it behind environment variables until it's ready.

**Comment by Mrmaxmeier** [comment]
Discarding all data that isn't visible to the user sounds like it's the best option considering backwards (and forwards) compatibility. Adding `_fetched_for` might be useful later though.

There are a few options for triggering the gc:
- on startup
- on shutdown
- every `x` minutes of uptime
- after `x` db operations

what do you think?
`requestIdleCallback` could be handy if it works in this situation.

**Comment by fasterthanlime** [comment]
> requestIdleCallback could be handy if it works in this situation.

I didn't know about it! ([docs page](https://developers.google.com/web/updates/2015/08/using-requestidlecallback?hl=en)), sounds interesting

I was thinking of triggering the GC on tab closes & startup

**Comment by fasterthanlime** [comment]
@Mrmaxmeier finally merged this, already works quite well! Will make a few changes I think.

**Pull request merged**